### PR TITLE
Remove reservoirState from BlackoilModelEbos

### PR DESCRIPTION
**Pull request by totto82**
**Motivation** 
Currently the solution state is stored both as global arrays in the BlackoilState and as local solution variables  for each cell in ebos and great care need to be taken to communicate the solution between these two storages. The aim of this PR is to avoid this duplication. This PR does not remove the BlackoilState completely from flow_ebos, it is still kept in the signature of the step methods in AdaptiveTimeStepping and NonlinearSolver. And also in the initialization, for restart and for output of the initial state. After the initial step, an empty reservoirState is passed around in the code to avoid abuse. 

**Changes**
1) Use the solution variable directly in RelativeChange(...)
2) Add a method in the RateConverter that takes the simulator instead of the state.
3) Pass the reservoir pressure directly to the well initialization.
4) Move convertInput(...) to SimulatorFullyImplicitBlackoilEbos.hpp.
This code is only used to convert the initial reservoir state.
5) Modify  updateState(...). The solution variable is updated directly and adaptPrimaryVariable(...)
from ewoms is used to switch primary variables. An epsilon is passed to adaptPrimaryVarible(...) after a switch of primary variables to make it harder to immediately switch back.

**Dependencies**
OPM/opm-core#1172 and OPM/ewoms#215

**Testing** 
flow_ebos is tested on SPE9, Norne, Model2-0, Model2-5
flow_ebos_solvent is tested on SPE5CASE1

This PR introduces small changes in the variable switching code that changes the convergence behavior and therefore also the results. The conditional usage of an epsilon just after a switch (not always as the old code) to avoid oscillations is meant to improve the convergence. Taking into account VAPPARS in the variable switching makes fixes a bug in the old code. The impact of VAPPARS in the variable switching is however very small for the only case with currently run with VAPPARS i.e. The Norne case.  

I have verified that this PR does not deteriorate the results and performance of the simulator for the tested cases.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this ewoms=215 opm-core=1172 please

**Comment by totto82** [comment]
Update testes:  
flow_ebos_2p on SPE1CASE2_2P 
flow_mpi and flow_ebos on Norne with mpi.

**Comment by totto82** [comment]
jenkins build this ewoms=215 opm-core=1172 please

**Comment by totto82** [comment]
Some difference in the results are expected since this PR changes the primary variable switching code. i.e. the test failure for SPE1, SPE3 and SPE9. The deviations are however all small. Good match is verified on all the tested models (see comments above) 
This is ready to be reviewed and merged from my side.

File: opm/autodiff/RateConverter.hpp [hunk 1]
```diff
@@ -450,7 +450,62 @@ namespace Opm {
...
+                auto& ra = attr_.attributes(0);
+                auto& p  = ra.pressure;
+                auto& T  = ra.temperature;
+                std::size_t n = 0;
+
+                ElementContext elemCtx( simulator );
+                const auto& gridView = simulator.gridView();
+
+                const auto& elemEndIt = gridView.template end</*codim=*/0>();
+                for (auto elemIt = gridView.template begin</*codim=*/0>();
```
Comment (atgeirr) [inline opm/autodiff/RateConverter.hpp]: I may be wrong, but does not this loop iterate over all cells present in the partition, and not just those *owned by* the partition? If so, the counts and averages will be slightly wrong, as overlap cells are considered multiple times. I believe a test like ` if (elem.partitionType() != Dune::InteriorEntity)` would resolve this?

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 2]
```diff
@@ -135,8 +137,7 @@ class SimulatorFullyImplicitBlackoilEbos
     /// \param[in,out] timer       governs the requested reporting timesteps
     /// \param[in,out] state       state of reservoir: pressure, fluxes
     /// \return                    simulation report, with timing data
-    SimulatorReport run(SimulatorTimer& timer,
-                        ReservoirState& state)
+    SimulatorReport run(SimulatorTimer& timer,  ReservoirState& state)
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Did you mean to remove the argument? I don't see any change here?

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 3]
```diff
@@ -146,11 +147,21 @@ class SimulatorFullyImplicitBlackoilEbos
...
 
+        // communicate the initial solution to ebos
+        if (timer.initialStep()) {
+            convertInput(/*iterationIdx=*/0, state, ebosSimulator_ );
+            ebosSimulator_.model().invalidateIntensiveQuantitiesCache(/*timeIdx=*/0);
+        }
+
         if (output_writer_.isRestart()) {
             // This is a restart, populate WellState and ReservoirState state objects from restart file
+            ReservoirState state(grid().size(0), grid().numFaces(), phaseUsage_.num_phases);
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Why create this object and not use the already existing (but now shadowed) `state`? If you really want a new one, the one used inside this scope should have a different name.

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 4]
```diff
@@ -146,11 +147,21 @@ class SimulatorFullyImplicitBlackoilEbos
...
             output_writer_.initFromRestartFile(phaseUsage_, grid(), state, prev_well_state, extra);
-            initHydroCarbonState(state, phaseUsage_, Opm::UgGridHelpers::numCells(grid()), has_disgas_, has_vapoil_);
+            initHydroCarbonState(state, phaseUsage_, Opm::UgGridHelpers::numCells(grid()), has_disgas_, has_vapoil_);
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Whitespace fubar?

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 5]
```diff
@@ -198,6 +209,11 @@ class SimulatorFullyImplicitBlackoilEbos
                                     prev_well_state,
                                     restorefilename,
                                     desiredRestoreStep );
+            initHydroCarbonState(state, phaseUsage_, Opm::UgGridHelpers::numCells(grid()), has_disgas_, has_vapoil_);
+            initHysteresisParams(state);
+            // communicate the restart solution to ebos
+            convertInput(0, state, ebosSimulator_);
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: The convertInput() and cache invalidation calls seem to be repeated. Would it be wrong to collect them below?

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 6]
```diff
@@ -854,6 +900,104 @@ class SimulatorFullyImplicitBlackoilEbos
...
+                }
+            } else {
+                // for oil-water case oil pressure should be used as primary variable
+                cellPv[BlackoilIndices::pressureSwitchIdx] = oilPressure[cellIdx];
+            }
+        }
+
+        if( iterationIdx == 0 )
+        {
+            simulator.model().solution( 1 /* timeIdx */ ) = solution;
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: What does this do, why is it necessary?

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 7]
```diff
@@ -383,47 +383,90 @@ namespace Opm {
...
+        double relativeChange(const Dummy&, const Dummy&) const
         {
-            std::vector< double > p0  ( previous.pressure() );
-            std::vector< double > sat0( previous.saturation() );
+            Scalar resultDelta = 0.0;
+            Scalar resultDenom = 0.0;
 
-            const std::size_t pSize = p0.size();
-            const std::size_t satSize = sat0.size();
+	    const auto& elemMapper = ebosSimulator_.model().elementMapper();
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Tab used instead of spaces.

File: opm/autodiff/ParallelDebugOutput.hpp [hunk 8]
```diff
@@ -413,13 +400,9 @@ namespace Opm
...
                     const auto& data = pair.second.data;
-                    const size_t stride = data.size()/localState_.numCells();
 
-                    for( size_t i=0; i<stride; ++i )
-                    {
-                        // write all data from local state to buffer
-                        write( buffer, localIndexMap_, data, i, stride );
-                    }
+                    // write all data from local data to buffer
+                    write( buffer, localIndexMap_, data);
```
Comment (totto82) [inline opm/autodiff/ParallelDebugOutput.hpp]: The data we write is `data::Solution` from opm-output where "stide" is not used. I think this is just some leftover from the time the reservoirState was written out directly.

File: opm/autodiff/RateConverter.hpp (hunk 1 above)
Comment (totto82) [inline opm/autodiff/RateConverter.hpp]: I think you are right. Good catch.

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp (hunk 3 above)
Comment (totto82) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Just some leftovers from a experimental code patch....

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp (hunk 5 above)
Comment (totto82) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Not wrong. I just wanted to make it very clear that  the simulator
1) gets a state either from the initialization or one of the restart approaches
2) convert the state to the solution state in ebos
3) clear the cache in ebos.

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp (hunk 6 above)
Comment (totto82) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: This stores the solution in the beginning av the time step ( t^(n-1) . )

**Comment by totto82** [comment]
jenkins build this ewoms=215 please

**Comment by totto82** [comment]
jenkins build this ewoms=215 update_data please

**Comment by atgeirr** [comment]
jenkins build this ewoms=215 opm-data=238 please

**Comment by atgeirr** [comment]
Looking at the test failures (that makes us require an opm-data update), I wonder about this test:

https://ci.opm-project.org/job/opm-simulators-PR-builder/549/testReport/junit/(root)/mpi/compareECLFiles_flow_ebos_SPE3CASE1/

The FOPR and FOPT changes (for example) are dramatic (from nothing to something). Have you looked at this?

**Comment by totto82** [comment]
> The FOPR and FOPT changes (for example) are dramatic (from nothing to something). Have you looked at this?

With this patch SPE1CASE1 doesn't fail the first step. Since the number of time step are different all comparisons are wrong. The results are identical in the eyeball norm.

**Comment by atgeirr** [comment]
Why does Jenkins fail here (the SPE9 group case fails), can it have something to do with the recent group control troubles @GitPaean?

**Comment by totto82** [comment]
> Why does Jenkins fail here (the SPE9 group case fails), can it have something to do with the recent group control troubles @GitPaean?

Only small differences for SPE9. I guess it's just the numerical noise.

**Comment by GitPaean** [comment]
> Only small differences for SPE9. I guess it's just the numerical noise.

The difference is small and nothing wrong with the result. 
```
27: Failure for keyword PRESSURE
27: 	 Fails for 6 entries
27: 	 Largest absolute deviation: 1.628420e-01
27: 	 Largest relative deviation: 4.524050e-05
```
I think the PR in `opm-data` needs to be regenerated or updated to fix the failure.

**Comment by atgeirr** [comment]
The question is why the opm-data PR did not contain exactly the needed values to be ok. I'll try to regenerate and see what happens.

**Comment by atgeirr** [comment]
> The question is why the opm-data PR did not contain exactly the needed values to be ok. I'll try to regenerate and see what happens.

Actually, we cannot do that, right? Because it would use the same branch etc. Need to find another solution.

**Comment by GitPaean** [comment]
The PR in `opm-data` looks like does not contain things related to `SPE9_CP_GROUP`.  Maybe there is something missing there?

**Comment by totto82** [comment]
I just noticed the same. Maybe @akva2 needs to do some voodoo for Jenkins to understand that the spe9_group should be updated as well.

**Comment by akva2** [comment]
just remember to use the old PR before you regen and it will work fine.

but yes, as totto82 says, when a new test is added, the updater script needs to be updated. 10 mins.

**Comment by akva2** [comment]
time estimate was not adjusted for working on a laptop and a couple of details, but anyways, https://github.com/OPM/opm-simulators/pull/1256

**Comment by atgeirr** [comment]
> time estimate was not adjusted for working on a laptop and a couple of details, but anyways, #1256

Thanks a lot!

**Comment by atgeirr** [comment]
jenkins build this ewoms=215 update_data please

**Comment by akva2** [comment]
jenkins build this ewoms=215 opm-data=239 please

**Comment by atgeirr** [comment]
All in!

File: opm/autodiff/RateConverter.hpp [hunk 9]
```diff
@@ -450,7 +450,66 @@ namespace Opm {
...
+             *                 it wraps a ParallelISTLInformation. Parameter
+             *                 is optional.
+             */
+            template <typename ElementContext, class EbosSimulator>
+            void defineState(const EbosSimulator& simulator)
+            {
+
+                //const int numCells = cellPvtIdx_.size();
+                //const Region region = std::vector<int>(numCells, 0);
+                auto& ra = attr_.attributes(0);
```
Comment (blattms) [inline opm/autodiff/RateConverter.hpp]: I have been looking at this code for a while. I still cannot see how the new function `RateConverter::defineState(EbosSimulator)` calculates averages over the regions. For me it seems like the average is done once over the whole domain. What am I missing?

Comment (GitPaean) [reply opm/autodiff/RateConverter.hpp]: I might remember wrong. I am not sure we support averages over separate regions.

Comment (GitPaean) [reply opm/autodiff/RateConverter.hpp]: I got the impression form the following code from StandardWellsDense_impl.hpp. (line number is different from master.)
```
787                 // the average hydrocarbon conditions of the whole field will be used
788                 const int fipreg = 0; // Not considering FIP for the moment.
789 
790                 rate_converter_.calcCoeff(well_rates, fipreg, convert_coeff);
791                 well_voidage_rates[w] = std::inner_product(well_rates.begin(), well_rates.end(),
792                                                            convert_coeff.begin(), 0.0)
```

Comment (totto82) [reply opm/autodiff/RateConverter.hpp]: You are right. It doesn't. We only support calculating averages over the whole domain. If option 13 is set in WELSPECS. The averages are supposed to be calculated for each FIP region. I am currently debugging the RateConverter.hpp. There are several things in this code that may not be correct.

**Pull request merged**